zjit/src/state.rs

```diff
@@ -219,3 +239,139 @@ pub extern "C" fn rb_zjit_assert_compiles(_ec: EcPtr, _self: VALUE) -> VALUE {
     ZJITState::enable_assert_compiles();
     Qnil
 }
+
+/// Call `rb_profile_frames` and format the result into buffers.
+fn record_profiling_frames() -> (i32, Vec<VALUE>, Vec<i32>) {
+    // Use the same buffer size as Stackprof.
+    const BUFF_LEN: usize = 2048;
+
+    let mut frames_buffer = vec![VALUE(0_usize); BUFF_LEN];
+    let mut lines_buffer = vec![0; BUFF_LEN];
+
+    let stack_length = unsafe {
+        rb_profile_frames(
+            0,
+            BUFF_LEN as i32,
+            frames_buffer.as_mut_ptr(),
+            lines_buffer.as_mut_ptr(),
+        )
+    };
+
+    // Trim at `stack_length` since anything past it is redundant
+    frames_buffer.truncate(stack_length as usize);
+    lines_buffer.truncate(stack_length as usize);
+
+    (stack_length, frames_buffer, lines_buffer)
+}
+
+/// Record a backtrace with ZJIT side exits
+#[unsafe(no_mangle)]
+pub extern "C" fn rb_zjit_record_exit_stack(_exit_pc: *const VALUE)
```

Comment:
I think `_exit_pc` should be renamed to `exit_pc`.

Also, it might be too much, but I'm tempted to share this under `jit/` too, given how big these functions are and the fact that they likely don't need YJIT/ZJIT-specific logic in the aggregation itself?